resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryBudgetImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ *
+ *  Copyright 2018 Jan Sykora at GoodData(R) Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+
+package io.github.resilience4j.retry.internal;
+
+import io.github.resilience4j.bitset.RingBitSet;
+import io.github.resilience4j.retry.Retry;
+import io.github.resilience4j.retry.RetryConfig;
+import io.github.resilience4j.retry.RetryMetrics;
+import io.github.resilience4j.retry.event.RetryEvent;
+import io.github.resilience4j.retry.event.RetryOnErrorEvent;
+import io.github.resilience4j.retry.event.RetryOnIgnoredErrorEvent;
+import io.github.resilience4j.retry.event.RetryOnSuccessEvent;
+import io.vavr.CheckedConsumer;
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class RetryBudgetImpl implements Retry {
+
+    private final RetryMetrics metrics;
+    private final RetryEventProcessor eventProcessor;
+
+    private String name;
+    private RetryConfig config;
+    private int maxAttempts;
+    private Function<Integer, Long> intervalFunction;
+    private Predicate<Throwable> exceptionPredicate;
+    private int bufferSize;
+    // to avoid integer division
+    private double _bufferSize;
+    private double retryThreshold;
+    private LongAdder succeededAfterRetryCounter;
+    private LongAdder failedAfterRetryCounter;
+    private LongAdder succeededWithoutRetryCounter;
+    private LongAdder failedWithoutRetryCounter;
+    private CheckedConsumer<Long> sleepFunction;
+    private RingBitSet ringBitSet;
+
+    public RetryBudgetImpl(String name, RetryConfig config) {
+        this.name = name;
+        this.config = config;
+        this.maxAttempts = config.getMaxAttempts();
+        this.intervalFunction = config.getIntervalFunction();
+        this.exceptionPredicate = config.getExceptionPredicate();
+        this.bufferSize = config.getBufferSize();
+        this._bufferSize = config.getBufferSize();
```

Comment:
Let's find a better name for this variable

Reply:
I will keep only `bufferSize` and internally it will be double